### DISCUSSION TITLE
WS2812 json

### DISCUSSION THREAD

**Discussion by sorriso93** [RavenSystem/esp-homekit-devices#1709] [MEPLHAA Scripts]
Hello
I'm trying to do a simple strip controller for my WS2812 using and electrodragon but I'm stuck with the json.
Tried this but I don't know where it is the error (I get json invalid or a freeze on boot)

{"c":{"q":850,"n":"STRSCRIVMAX","o":1},"a":[{"t":30,"ty":8,"g":[2,1,192]},{"i":0.5,"1": {"s":[{"a":1}]}}]}

**Comment by RavenSystem** [comment]
Remove "q":850 because you are not using PWM.

**Comment by sorriso93** [comment]
Hello Jose, I removed it. All is good except I get green and blue inverted; I tried almost all nrz configuration suggested in the wiki with no success, brightness control is ok, red color is ok but blue and green are inverted. I cannot find any other setting to be changed in the json except the nrz for WS2812, WS2811, WS2812B but I fear having tried all that is not related... any suggestion?
I'm using a WS2812B 12v and electrodragon led strip hardware.

in the debug I see if I set green (the strip becomes blue)
Jan 01 00:07:27 <1> Lightbulb setter ON 1, BRI 59
Jan 01 00:07:27 <1> Lightbulb setter HUE 119, SAT 88
Jan 01 00:07:27 New chromaticity: [0.153647, 0.694449]
Jan 01 00:07:27 <1> Target Color = 196, 38665, 0, 1424, 0
Jan 01 00:07:27 <1> Exec Action 1
Jan 01 00:07:27 [3] Send Ev
Jan 01 00:07:28 [5] Send Ev
Jan 01 00:07:28 Color fixed
Jan 01 00:07:32 Saving states

in the debug I see if I set blue (the strip becomes green):
Jan 01 00:08:54 <1> Lightbulb setter ON 1, BRI 59
Jan 01 00:08:54 <1> Lightbulb setter HUE 236, SAT 100
Jan 01 00:08:54 New chromaticity: [0.135054, 0.0656584]
Jan 01 00:08:54 <1> Target Color = 0, 207, 38665, 0, 0
Jan 01 00:08:54 Color fixed
Jan 01 00:08:54 <1> Exec Action 1
Jan 01 00:08:54 [3] Send Ev
Jan 01 00:08:54 [5] Send Ev
Jan 01 00:08:59 Saving states

in the debug I see if I set red (and the strip is red :-))
Jan 01 00:09:42 <1> Lightbulb setter ON 1, BRI 59
Jan 01 00:09:42 <1> Lightbulb setter HUE 354, SAT 100
Jan 01 00:09:42 New chromaticity: [0.686188, 0.305402]
Jan 01 00:09:42 <1> Target Color = 38665, 0, 365, 0, 0
Jan 01 00:09:42 Color fixed
Jan 01 00:09:42 <1> Exec Action 1
Jan 01 00:09:42 [3] Send Ev
Jan 01 00:09:42 [5] Send Ev
Jan 01 00:09:47 Saving states

**Reply by RavenSystem** [reply]
Add undocumented feature "color map": ``"cm":[0,1,2,3,4]`

```json
{"c":{"n":"STRSCRIVMAX","o":1},"a":[{"t":30,"ty":8,"cm":[0,1,2,3,4],"g":[2,1,192]},{"i":0.5,"1": {"s":[{"a":1}]}}]}
```

**Reply by sorriso93** [reply]
and the default is 0? so I have to try the others ;-)

**Reply by RavenSystem** [reply]
Default is `"cm":[1,0,2,3,4]`

**Reply by sorriso93** [reply]
> Add undocumented feature "color map": ``"cm":[0,1,2,3,4]`
> 
> ```json
> {"c":{"n":"STRSCRIVMAX","o":1},"a":[{"t":30,"ty":8,"cm":[0,1,2,3,4],"g":[2,1,192]},{"i":0.5,"1": {"s":[{"a":1}]}}]}
> ```

green > red, blue >> green, red >> blue
how is related the mapping with RGB?

**Reply by RavenSystem** [reply]
{"c":{"n":"STRSCRIVMAX","o":1},"a":[{"t":30,"ty":8,"cm":[1,2,0,3,4],"g":[2,1,192]},{"i":0.5,"1": {"s":[{"a":1}]}}]}

**Reply by sorriso93** [reply]
> {"c":{"n":"STRSCRIVMAX","o":1},"a":[{"t":30,"ty":8,"cm":[1,2,0,3,4],"g":[2,1,192]},{"i":0.5,"1": {"s":[{"a":1}]}}]}

I was playing also myself, tried 21043 but was not good, red and blue inverted, as yours
tried 12304, strange behaviour on red (brightness goes to zero and blue and green inverted)
tried 12340 also here strange behaviour. blue >> red, green >> blue, red >> low brightness red...

**Reply by RavenSystem** [reply]
Did you try? {"c":{"n":"STRSCRIVMAX","o":1},"a":[{"t":30,"ty":8,"cm":[1,2,0,3,4],"g":[2,1,192]},{"i":0.5,"1": {"s":[{"a":1}]}}]}

**Reply by RavenSystem** [reply]
Don't change 3,4 in any test, and leave them at last. They are for cool and warm white.

**Reply by sorriso93** [reply]
> {"c":{"n":"STRSCRIVMAX","o":1},"a":[{"t":30,"ty":8,"cm":[1,2,0,3,4],"g":[2,1,192]},{"i":0.5,"1": {"s":[{"a":1}]}}]}

yes
red >> green, blue >> red, green >> blue

**Reply by RavenSystem** [reply]
0 is Red.
1 is Green.
2 is Blue.

And "cm" array has the order of colors.

**Reply by sorriso93** [reply]
> > {"c":{"n":"STRSCRIVMAX","o":1},"a":[{"t":30,"ty":8,"cm":[1,2,0,3,4],"g":[2,1,192]},{"i":0.5,"1": {"s":[{"a":1}]}}]}
> 
> yes red >> green, blue >> red, green >> blue

Found the correct combination, "cm":[2,0,1,3,4]
Many thanks for your support!

**Reply by RavenSystem** [reply]
Great!
